Enhaced RDoc for CSV::filter

BurdetteLamar opened the pull request:
Better RDoc for this versatile and powerful method.

BurdetteLamar commented:
Marking as draft.  Still a couple of things to do.

BurdetteLamar commented:
Now ready for review and merge.

kou commented on lib/csv.rb:
```diff
@@ -1006,60 +1006,149 @@ def instance(data = $stdout, **options)
+    # \String input without headers:
+    #
+    #   in_string = "foo,0\nbar,1\nbaz,2"
+    #   CSV::filter(in_string) {|row| row[0].upcase!; row[1] = - row[1].to_i }
```
Why do you use multiple sentences in a line by `;`?
It may be difficult to read.

BurdetteLamar replied:
Fixed in ten places.

kou commented on lib/csv.rb:
```diff
@@ -1006,60 +1006,149 @@ def instance(data = $stdout, **options)
+    #   File.read('t.csv') # => "Name,Value\nFOO,0\nBAR,-1\nBAZ,-2\n"
+    #
+    # When neither +in_string_or_io+ nor +out_string_or_io+ given,
+    # parses from {ARGF}[https://ruby-doc.org/core/ARGF.html]
```
Could you use https://docs.ruby-lang.org/en/ instead of https://ruby-doc.org/ ?

BurdetteLamar replied:
Done.

kou commented on lib/csv.rb:
```diff
@@ -1006,60 +1006,149 @@ def instance(data = $stdout, **options)
+    #   # Put Ruby code into a file.
+    #   ruby = <<-EOT
+    #     require 'csv'
+    #     CSV::filter {|row| row[0].upcase!; row[1] = - row[1].to_i }
```
Could you use `CSV.filter` style instead of `CSV::filter` style?

BurdetteLamar replied:
Done.

kou commented on lib/csv.rb:
```diff
@@ -1006,60 +1006,149 @@ def instance(data = $stdout, **options)
+    #   # Put some CSV into a file.
+    #   File.write('t.csv', "foo,0\nbar,1\nbaz,2")
+    #   # Run the Ruby code with CSV filename as argument.
+    #   `ruby t.rb t.csv`
```
Umm. This is a bit tricky to show `ARGF` behavior...

At least, we should use `system` instead of `` `...` ``:

```suggestion
    #   system(Gem.ruby, "t.rb", "t.csv")
```

BurdetteLamar replied:
Done.

kou commented on lib/csv.rb:
```diff
@@ -1006,60 +1006,149 @@ def instance(data = $stdout, **options)
+    #   # Put Ruby code into a file.
+    #   ruby = <<-EOT
+    #     require 'csv'
+    #     p CSV::filter(headers: true) {|row| row[0].upcase!; row[1] = - row[1].to_i }
```
```suggestion
    #     CSV::filter(headers: true) {|row| row[0].upcase!; row[1] = - row[1].to_i }
```

BurdetteLamar replied:
Oops. Fixed.

BurdetteLamar commented:
Also:
- Added text at top to specify return value (Array of Arrays or CSV::Table) and passed row (Array or CSV::Row).
- Added return value in code examples, to emphasize difference between headers and not.

kou commented:
Thanks!
